### ISSUE TITLE
Hexagon layout export to neural networks (addressing/input formats)

### ISSUE THREAD

**Issue by rustyconover** [uber/h3#335]
Hi,

I've been thinking about how to utilize a collection of H3 hexagons with a neural network and I have questions about the best/right way to present the H3 data to the neural network while retaining the spacial relationship between hexagons.

Passing image with rectangular pixels to a neural network is a well travelled road filled with examples, but obviously hexagons aren't squares.

It might be best to start with an example.

## Temperature Example

I've created a collection of H3 index values associated with temperature at various H3 resolutions. 

You can see the example here:

https://observablehq.com/@rustyconover/2020-04-22-apparent-temperature-visualized-with-h3-leaflet

I'd like to use those hexagon based temperatures as inputs into a neural network to perform some regression tasks.  One example would be predicting the temperature of a missing hexagon.

Looking at the Uber's presentations online such as:

https://portal.opengeospatial.org/files/?artifact_id=90338

There seems to be a suggestion to render the data to a square matrix using `h3.experimentalH3ToLocalIj`.  So applying that to my example yields:

https://observablehq.com/@rustyconover/example-h3-index-to-local-i-j-coordinates-with-h3-js

Using the local `i,j` coordinates seem to have limitations at the larger resolutions for areas like the continental USA.  Is there another approach I should consider for addressing while presenting the data to the neural network?

## Uber's Ludwig

Recently Uber released [Ludwig](https://uber.github.io/ludwig/) with some H3 feature support.

Looking over the documentation of Ludwig's [H3 features](https://uber.github.io/ludwig/user_guide/#h3-features):

> This encoder encodes each components of the H3 representation (mode, edge, resolution, base cell and childern[sic] cells) with embeddings. Chidren cells with value 0 will be masked out. After the embedding, all embeddings are summed and optionally passed through a stack of fully connected layers. 

Ludwig appears to be able to handle features with H3 indexes of different resolutions and reduces the cardinality through embeddings, but it doesn't appear to use any hexagon based convolutions.

## Additional Links

It seems there have bee some recent open source code releases featuring hexagon based convolution:  

### Hexagdly 

https://github.com/ai4iacts/hexagdly for pytorch.

Converting from a collection of H3 indexes to their expected addressing scheme to be a bit more complicated than expected.

https://github.com/ai4iacts/hexagdly/blob/master/notebooks/how_to_apply_adressing_scheme.ipynb

Am I missing the easy happy path API call to extract a collection of H3 indexes such that it is easily exported to a neural network via an ordered matrix?

Many thanks,

Rusty

**Comment by dfellis** [comment]
So I don't think there's a perfect answer for this right now.

[The `experimentalH3ToLocalIj` function](https://h3geo.org/docs/api/traversal) lets you convert to a 2D coordinate system (IJ where J is skewed, similar [to this diagram of axial coordinates](https://www.redblobgames.com/grids/hexagons/#coordinates-axial) but the orientation depends on the icosahedron face you're on), but doesn't work across icosahedron boundaries (hence "local").

This gets you a classic grid that you can work with existing pixel-oriented NNs, though with some distortion that hopefully the NN figures out.

But it doesn't get you the same neighbor relations as you'd like (that you have 6 equal neighbors instead of 4 edge neighbors and 4 vertex neighbors). You can construct a map of those relations by [calling `getH3UnidirectionalEdgesFromHexagon` on your entire set of H3Indexes](https://h3geo.org/docs/api/uniedge) and turning that map (plus the original hexagon) into the net convolution for the next layer, making the interconnectedness more available (and you could repeat this convolution a few times to effectively increase the "k-ring" applied). The order of the output from that function is guaranteed to be consistent between hexagons on the same icosahedron face, so you could also use that information to keep some directional biases along with the neighboring biases, though off the top of my head I'm not sure how.

You could combine these two, since the actual hex neighbors are going to be a subset of the 8 IJ "neighbors" and you could apply a convolution matrix with that pattern maintained, [which a really bright guy I know at Uber did a couple of years ago :)](https://www.youtube.com/watch?v=z3PaGIQTFSE&list=PLLEUtp5eGr7CNf9Bj3w3i30rzaU8lKZeV&index=14&t=0s). That can then be the basis for the layering.

I hope that helps?

**Comment by nrabinowitz** [comment]
Closing this for now, as @dfellis's answer gives the state of the art at the moment.
